zrml/parimutuel/src/lib.rs

```diff
@@ -0,0 +1,467 @@
+            // this allows us to query all outstanding shares for each discrete asset
+            let outcome_total = T::AssetManager::total_issuance(winning_asset);
+            // if there are no outstanding reward shares, but the pot account is not empty
+            // then use the refund extrinsic to get the initial bet back
+            ensure!(outcome_total != BalanceOf::<T>::zero(), Error::<T>::NoRewardShareOutstanding);
+            let winning_balance = T::AssetManager::free_balance(winning_asset, &who);
+            ensure!(!winning_balance.is_zero(), Error::<T>::NoWinningShares);
+            debug_assert!(
+                outcome_total >= winning_balance,
+                "The outcome issuance should be at least as high as the individual balance of \
+                 this outcome!"
+            );
+
+            let pot_total = T::AssetManager::free_balance(market.base_asset, &pot_account);
+            let payoff_ratio_mul_base: BalanceOf<T> =
+                bdiv_floor(pot_total.saturated_into(), outcome_total.saturated_into())?
+                    .saturated_into();
+            let payoff: BalanceOf<T> = bmul_floor(
+                payoff_ratio_mul_base.saturated_into(),
+                winning_balance.saturated_into(),
+            )?
+            .saturated_into();
+
+            Self::check_values(
+                winning_balance,
+                pot_total,
+                outcome_total,
+                payoff_ratio_mul_base,
+                payoff,
+            )?;
+
+            let slashable_asset_balance = winning_balance;
+
+            T::AssetManager::slash(winning_asset, &who, slashable_asset_balance);
+
+            let remaining_bal = T::AssetManager::free_balance(market.base_asset, &pot_account);
+            let actual_payoff = payoff.min(remaining_bal);
+
+            T::AssetManager::transfer(market.base_asset, &pot_account, &who, actual_payoff)?;
+
+            Self::deposit_event(Event::RewardsClaimed {
+                market_id,
+                asset: winning_asset,
+                balance: slashable_asset_balance,
+                actual_payoff,
+                sender: who.clone(),
+            });
+
+            Ok(())
+        }
+
+        fn do_claim_refunds(who: T::AccountId, refund_asset: AssetOf<T>) -> DispatchResult {
+            let market_id = match refund_asset {
+                Asset::ParimutuelShare(market_id, _) => market_id,
+                _ => return Err(Error::<T>::NotParimutuelOutcome.into()),
+            };
+            let market = T::MarketCommons::market(&market_id)?;
+            ensure!(market.status == MarketStatus::Resolved, Error::<T>::MarketIsNotResolvedYet);
+            ensure!(market.scoring_rule == ScoringRule::Parimutuel, Error::<T>::InvalidScoringRule);
+            ensure!(
+                matches!(market.market_type, MarketType::Categorical(_)),
+                Error::<T>::NotCategorical
+            );
```

Comment:
Shared code with `do_claim_rewards`, could be moved into `ensure_perimutual_market_resolved` for example.

Reply:
1e80853dfffa47ae3532457d2362b6f8ea640614